### ISSUE TITLE
Ability to edit existing image occlusion notes

### ISSUE THREAD

**Issue by Glutanimate** [Glutanimate/image-occlusion-enhanced#1]
Very difficult to implement with how I/O currently works. As each I/O yields in several different notes, we would first have to find a way to match all notes belonging to one image occlusion. This would be much easier if #2 was implemented.

Even if we find a way to associate all respective notes we would still have to figure out a way to handle changes to the SVG masks. It should be easy enough to update all other textual information, but how do we recognize a modified shape as such? How to preserve the scheduling information for each card? How do we handle new occlusions being added? etc...

**Comment by Glutanimate** [comment]
Another option would be to create a set of new I/O cards and reschedule them based on the mean ease and interval of the old cards.

As of 196f7086f1ea579c96915b9deb22bd03cf3ab89b, I/O 2 Enhanced is already capable of preserving SVG masks, so it would just come down to figuring out out how to implement this rescheduling.

**Comment by tmbb** [comment]
The right way is probably to pay some attention to SVG element id fields. I think svg-edit preserves the ids when you edit the document. If you store the id of the mask somehow, you'll be able to match the new cards to the old ones. This is all conjecture, I haven't touched this code in ages.

**Comment by Glutanimate** [comment]
That would probably be a good way to go about this. Though I'd have to make it obvious to the user that each mask element represents a card alongside all of its scheduling information. Otherwise they might do something like move an existing element to cover up a different piece of information which would lead to confusion when that "new" occlusion would suddenly appear as an old card. Maybe one way to do this would be to add some kind of temporary labels to the SVG signifying the interval of each mask, but I'll have to think about this again when I have some more time on my hands.

**Comment by tmbb** [comment]
Another way I've thougt about is to consider two masks equal even if they have the same id. This can be done by parsing the svg code of the mask (It's just xml, you can use the already included ElementTree library), deleting the id field and rendering it into a canonical form (normalizing whitespace, etc, I'm sure this is possible), but preserving all other attributes, including x and y coordinates and then you consider the masks,  and consider the elements equal if these canonical representations match. Alternatively (much harder), you can match the elements if the positions differ by a small amount, but this sounds like something that would open it's own can of worms.

**Comment by Glutanimate** [comment]
I think both approaches have their advantages and disadvantages. If you go by element IDs you have to make sure to communicate to the user that each shape represents a specific card. OTOH, if you match by geometric attributes you have to find a way to disregard smaller changes to the dimensions and positions of the shapes. 

I actually think of this last point this last point as an absolute necessity if we follow the geometric matching route. I don't see people having much use for this feature if they can't tweak existing shapes without losing their scheduling.

**Comment by Glutanimate** [comment]
Hi,
I've finally managed to implement this. Please feel free to give the new version of the add-on in the [dev-branch](https://github.com/Glutanimate/image-occlusion-enhanced/tree/dev) a spin. I ended up going with the first method you proposed, matching SVG IDs to a Note ID field. The new note generation code is based on [Simple Picture Occlusion](https://github.com/steveaw/anki_addons/blob/master/simple_picocc/simple_picocc_main.py) by @SteveAW, which I stumbled upon more or less by accident.

In any case, I will mark this issue as resolved for now, but I'd love to hear your feedback on the implementation!

**Comment by tmbb** [comment]
I'll look into it tomorrow! Thank you for improving the addon

**Comment by tmbb** [comment]
I haven't looked at the source code but I've tried it and it works great. Some thoughts:

- The names you've chosen for the kinds of note the addon generates (*Hide all, Reveal one*, etc) sound almost my confusing as my original names. I'd suggest changing to something like *Hide all, Guess one* or *Hide all, Ask for one*. On the other hand, this is entirely subjective and your names are much better than mine.
- If you can show a special *edit occlusion* button when the user chooses to add the note during the review, can't you also somehow monkey patch the "AnkiMainWindow.onEditcurrent()" method and show the SVG window immediately instead of showing the note editor? I have now idea how feasible this is, but it would be a nice improvement from the user's point of view. The problem I see is that it might break in future Anki version or something. Monkey patching Anki's methods is always risky in theory but it seems to work well in practice.
- I've been reading Anki's source and it seem to me that integrating Anki's editor into SVG-edit's window might not be as difficult as I thought it would be. to create a `QWidget` to encalsulate `Editor.outerLayout` and in principle, that widget would encapsulate the relevant parts of the editor. It would be mostly cosmetic, and I don't know if it would be worth it, but it certainly seems doable.

**Comment by Glutanimate** [comment]
Hi again and thanks for the feedback!

> The names you've chosen for the kinds of note the addon generates (Hide all, Reveal one, etc) sound almost my confusing as my original names...

Yep, I'm not perfectly happy with them myself. The tooltips on each button somewhat ameliorate this as I think they do a good job of explaining what each occlusion mode does, but I would still like for the labels to be clearer in the first place. Your suggestions are good, but I don't think they hit the perfect spot, either (especially if I consider the new third occlusion mode).


> If you can show a special edit occlusion button when the user chooses to add the note during the review, can't you also somehow monkey patch the "AnkiMainWindow.onEditcurrent()" method...

I actually [explored that possibility](https://github.com/Glutanimate/image-occlusion-enhanced/commit/27a129350dddd77f756e0308fe7bc4f00e690ae1#diff-5349acc68c554ca9e979302f07dab46eR304), but it was causing a number of very strange bugs in the reviewer, so I had to [revert the commit](https://github.com/Glutanimate/image-occlusion-enhanced/commit/a56a24a26f5f3348aa1e081419aec17e03e1e052). In the meantime I've also discovered that quite a few users actually do like to enter note-specific details, so I've more or less abandoned this idea, as appealing as it might be in most usage scenarios.

> I've been reading Anki's source and it seem to me that integrating Anki's editor into SVG-edit's window might not be as difficult as I thought it would be...

That's great to hear and I would love to finally do away with these ugly QPlainTextEdits. However - though I haven't looked into this properly yet - doesn't each Editor have a note assigned to it? I imagine that this might complicate things as we will usually have to create multiple notes. We would also have to hide a lot fields, e.g. the mask and ID fields, from the user.
